Gems/Atom/RHI/Code/Source/RHI/MultiDeviceQueryPool.cpp

```diff
@@ -0,0 +1,295 @@
+    {
+        return InitQuery(&query, 1);
+    }
+
+    ResultCode MultiDeviceQueryPool::InitQuery(MultiDeviceQuery** queries, uint32_t queryCount)
+    {
+        AZ_Assert(queries, "Null queries");
+        auto resultCode{ ResultCode::Success };
+
+        for (auto& [deviceIndex, deviceQueryPool] : m_deviceQueryPools)
+        {
+            AZStd::vector<RHI::Ptr<Query>> deviceQueries(queryCount);
+            AZStd::vector<Query*> rawDeviceQueries(queryCount);
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                deviceQueries[index] = RHI::Factory::Get().CreateQuery();
+                rawDeviceQueries[index] = deviceQueries[index].get();
+            }
+
+            resultCode = deviceQueryPool->InitQuery(rawDeviceQueries.data(), queryCount);
+
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                queries[index]->m_deviceQueries[deviceIndex] = deviceQueries[index];
+            }
+
+            if (resultCode != ResultCode::Success)
+            {
+                break;
+            }
+        }
+
+        if (resultCode == ResultCode::Success)
+        {
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                MultiDeviceResourcePool::InitResource(
+                    queries[index],
+                    []()
+                    {
+                        return ResultCode::Success;
+                    });
+            }
+        }
+
+        return resultCode;
+    }
+
+    ResultCode MultiDeviceQueryPool::ValidateQueries(MultiDeviceQuery** queries, uint32_t queryCount)
+    {
+        if (!queryCount)
+        {
+            AZ_Error("RHI", false, "MultiDeviceQuery count is 0");
+            return RHI::ResultCode::InvalidArgument;
+        }
+
+        for (uint32_t i = 0; i < queryCount; ++i)
+        {
+            auto* query = queries[i];
+            if (!query)
+            {
+                AZ_Error("RHI", false, "Null query %i", i);
+                return RHI::ResultCode::InvalidArgument;
+            }
+
+            if (query->GetQueryPool() != this)
+            {
+                AZ_Error("RHI", false, "MultiDeviceQuery does not belong to this pool");
+                return RHI::ResultCode::InvalidArgument;
+            }
+        }
+        return ResultCode::Success;
+    }
+
+    uint32_t MultiDeviceQueryPool::GetResultsCount(uint32_t queryCount, uint32_t deviceCount)
+    {
+        uint32_t perResultSize = m_descriptor.m_type == QueryType::PipelineStatistics
+            ? CountBitsSet(static_cast<uint64_t>(m_descriptor.m_pipelineStatisticsMask))
+            : 1;
+
+        if(deviceCount == 0)
+        {
+            deviceCount = RHISystemInterface::Get()->GetDeviceCount();
+        }
+
+        if(queryCount == 0)
+        {
+            queryCount = m_descriptor.m_queriesCount;
+        }
+
+        return perResultSize * queryCount * deviceCount;
```

Comment:
It is correct: The RHISystem has `N` devices that are indexed from `0` to `N-1`. So using the RHISystem's device count, we are always on the safe side. The bit masks on the other hand may have bits set for devices with an index `>= N` which will be ignored (e.g. `AllDevices`) simply has all bits set. So in your example if the RHISystem only reports one device, it will have index `0` and if the MD query pool is initialized with `1<<15` it will not have any actual SD query pools.

Reply:
You're right, in terms of memory ideally, we would use `msb(deviceMask & ((1 << rhiDeviceCount) - 1) + 1` as device count. However, in the general case (a single device, i.e., rhiDeviceCount=1) this doesn't change much besides the complexity of the code. Likely the memory overhead caused by this is negligible in most multi-device cases as well, so it might be premature optimization. I'd leave this as is for now and change when necessary. Anyway this API is still somewhat preliminary and we'll have to figure out what we actually need later.